pkg/pool-weighted/contracts/smart/LiquidityBootstrappingPool.sol

```diff
@@ -59,73 +88,315 @@ contract LiquidityBootstrappingPool is BaseWeightedPool, ReentrancyGuard {
-        _normalizedWeights = _normalizedWeights.insertUint64(weight, i * 64);
+    /**
+     * @dev Returns a fixed-point number representing how far along the current weight change is, where 0 means the
+     * change has not yet started, and FixedPoint.ONE means it has fully completed.
+     */
+    function _calculateWeightChangeProgress(bytes32 poolState) private view returns (uint256) {
+        uint256 currentTime = block.timestamp;
+        uint256 startTime = poolState.decodeUint32(_START_TIME_OFFSET);
+        uint256 endTime = poolState.decodeUint32(_END_TIME_OFFSET);
+
+        if (currentTime > endTime) {
+            return FixedPoint.ONE;
+        } else if (currentTime < startTime) {
+            return 0;
+        }
+
+        uint256 totalSeconds = endTime.sub(startTime);
+        uint256 secondsElapsed = currentTime.sub(startTime);
+
+        // In the degenerate case of a zero duration change, consider it completed (and avoid division by zero)
+        return totalSeconds == 0 ? FixedPoint.ONE : secondsElapsed.divDown(totalSeconds);
+    }
+
+    /**
+     * @dev When calling updateWeightsGradually again during an update, reset the start weights to the current weights,
+     * if necessary.
+     */
+    function _startGradualWeightChange(
+        uint256 startTime,
+        uint256 endTime,
+        uint256[] memory startWeights,
+        uint256[] memory endWeights
+    ) internal virtual {
+        bytes32 newPoolState = _poolState;
+
+        uint256 normalizedSum = 0;
+        for (uint256 i = 0; i < endWeights.length; i++) {
+            uint256 normalizedWeight = endWeights[i];
+            _require(normalizedWeight >= _MIN_WEIGHT, Errors.MIN_WEIGHT);
+
+            newPoolState = newPoolState.insertUint32(startWeights[i].compress32(), _START_WEIGHT_OFFSET + i * 32);
+            newPoolState = newPoolState.insertUint16(normalizedWeight.compress16(), _END_WEIGHT_OFFSET + i * 16);
+
+            normalizedSum = normalizedSum.add(normalizedWeight);
+        }
+        // Ensure that the normalized weights sum to ONE
+        _require(normalizedSum == FixedPoint.ONE, Errors.NORMALIZED_WEIGHT_INVARIANT);
+
+        // If we are initializing, start/end times are already 0
+        newPoolState = newPoolState.insertUint32(startTime, _START_TIME_OFFSET);
+        newPoolState = newPoolState.insertUint32(endTime, _END_TIME_OFFSET);
+
+        _poolState = newPoolState;
+
+        emit GradualWeightUpdateScheduled(startTime, endTime, startWeights, endWeights);
     }
 
-    // Assumes i is in range
-    function _getNormalizedWeight(uint8 i) private view returns (uint256) {
-        return _normalizedWeights.decodeUint64(i * 64);
+    function _interpolateWeight(
+        uint256 startWeight,
+        uint256 endWeight,
+        uint256 pctProgress
+    ) private pure returns (uint256 finalWeight) {
+        if (pctProgress == 0) return startWeight;
+        if (pctProgress == FixedPoint.ONE) return endWeight;
```

Comment:
```suggestion
        if (pctProgress >= FixedPoint.ONE) return endWeight;
```

Shouldn't be necessary, but will prevent ever returning values larger than `endWeight`.